### ISSUE TITLE
False positive for SC2248

### ISSUE THREAD

**Issue by LLyaudet** [koalaman/shellcheck#3405]
#### For bugs with existing features

- Rule Id (if any, e.g. SC1000): SC2248
- My shellcheck version (`shellcheck --version` or "online"): shellcheck (edge) v0.11.0+git57.29f0d8d
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

It seems https://www.shellcheck.net/ doesn't use --enable=all
But I checked that latest snap (latest/edge) mentions the same commit hash
as the latest commit on GitHub.

#### Here's a snippet or screenshot that shows the problem:

```bash
#!/usr/bin/env bash
# shellcheck enable=all
foo(){
  declare -i cd_result
  cd "$1" || {
    cd_result=$?;
    echo "foo no such directory";
    return ${cd_result};
  }
}
```

#### Here's what shellcheck currently says:
shellcheck --enable=all truc.sh 

In truc.sh line 7:
    return ${cd_result};
           ^----------^ SC2248 (style): Prefer double quoting even when variables don't contain special characters.

Did you mean:
    return "${cd_result}";

For more information:
  https://www.shellcheck.net/wiki/SC2248 -- Prefer double quoting even when v...

#### Here's what I wanted or expected to see:

I think the only valid exception to rule SC2248 is when the variable is declared as integer.
Script above is such a case,
and in my opinion there should be no warning in that case.

**Comment by LLyaudet** [comment]
Note that there is a slightly distinct case, where we return an integer via an indirection.
```
#!/usr/bin/env bash
# shellcheck enable=all
foo(){
  declare -r result_name="cd_result"
  declare -i "${result_name}"
  cd "$1" || {
    printf -v "${result_name}" "%d" $?
    echo "foo no such directory"
    return ${!result_name}
  }
}
```
In that case shellcheck doesn't return SC2248, but SC2086.
I don't know if both cases can be solved in one fix.
Otherwise if this first step is corrected one day, then maybe it will be ok to open another issue to solve the false positive for SC2086.

Well, it's strange I get also SC2248 when I test this sample script at:
https://www.shellcheck.net/

But the more complicated example in my code yields SC2086.

I've reproduced it:
```
#!/usr/bin/env bash
# shellcheck enable=all
foo(){
  declare -r result_name="cd_result_$1"
  declare -i "${result_name}"
  cd "$1" || {
    printf -v "${result_name}" "%d" $?
    echo "foo no such directory"
    return ${!result_name}
  }
}
```
With truly variable name using $1 in it, I can get SC2086 also at:
https://www.shellcheck.net/

Maybe having everytime SC2248 would be simpler to solve :).
